### PR TITLE
perf(scroll-dispatcher): lazily subscribe to global events

### LINKED ISSUE #3237: [Performance] ScrollDispatcher should subscribe lazily, and destroy subscriptions when they are no longer used
ScrollDispatcher needs to create subscriptions to the window scroll/resize events lazily (when they are actually needed by the UI) and destroy them once the UI using them gets destroyed.

Currently it simply subscribes to the scroll and resize events on init, which is when the page loads.
Since it is not utilizing the zone.runOutsideAngular functionality, then it means that every time a scroll or resize event happens, it will re-render the UI several times.

This usually goes unnoticed in smaller applications, but when you have bigger applications, this slows down the UI a lot.
It also removes performance control from the developers.

It's fine to emit the events when the dispatcher is actually being used by an overlay or whatnot, but it makes no sense to trigger zone to re-render the view every time the user scrolls, when there's nothing new to re-render.

I'd suggest making the subscriptions lazy and destroy them when components unsubscribe, or if possible using the zone.runOutsideAngular functionality to never trigger a re-render in the first place.

version 2.0.0-beta.2

### PR DESCRIPTION
**Pull request by crisbeto**
Switches the `ScrollDispatcher` to only subscribe to `scroll` and `resize` events if there any registered callbacks. Also unsubscribes once there are no more callbacks.

Fixes #3237.

### CODE REVIEW

File: src/lib/core/overlay/scroll/scroll-dispatcher.ts [hunk 1]
```diff
@@ -19,18 +20,15 @@ export class ScrollDispatcher {
   /** Subject for notifying that a registered scrollable reference element has been scrolled. */
   _scrolled: Subject<void> = new Subject<void>();
 
+  /** Keeps track of the global `scroll` and `resize` subscriptions. */
```
Comment (andrewseguin) [inline src/lib/core/overlay/scroll/scroll-dispatcher.ts]: Add comment about when this should have subscriptions (when there is at least one actively registered scrollable).

**Comment by andrewseguin** [comment]
Can you add tests for this? At minimum we should check that globalSubscriptions is null when there are no registered scrollables, and it exists when at least one scrollable is registered.

**Comment by crisbeto** [comment]
It feels weird having to make the subscription public only to be able to test it, but I should be able to use a spy to check whether stuff gets invoked.

**Comment by andrewseguin** [comment]
Just thought about the case of having no cdk-scrollables on the page but still wanting to respond to scroll events. We'll still want the tooltip to reposition in this case.

Does it make more sense to assign the global subscriptions when `scrolled` is called? It's invocation shows that there is intent from some component to listen to the scroll event.

**Comment by crisbeto** [comment]
I see your point @andrewseguin, but I'm not sure how we'd know to unbind the global handler once there are no more `scrolled` subscriptions.

**Comment by crisbeto** [comment]
Changed as discussed @andrewseguin. Could you take another look?

**Comment by QuentinFchx** [comment]
I might not understand all the stuff going behind the scene, but why are you not using RX operators such as [refCount](http://reactivex.io/documentation/operators/refcount.html) here?

File: src/lib/core/overlay/scroll/scroll-dispatcher.ts [hunk 2]
```diff
@@ -19,26 +20,28 @@ export class ScrollDispatcher {
...
   /**
    * Registers a Scrollable with the service and listens for its scrolled events. When the
    * scrollable is scrolled, the service emits the event in its scrolled observable.
    * @param scrollable Scrollable instance to be registered.
    */
   register(scrollable: Scrollable): void {
     const scrollSubscription = scrollable.elementScrolled().subscribe(() => this._notify());
+
     this.scrollableReferences.set(scrollable, scrollSubscription);
+    this._addGlobalSubscription();
```
Comment (andrewseguin) [inline src/lib/core/overlay/scroll/scroll-dispatcher.ts]: Should we still be adding/removing the global subscription in register and deregister? I thought that we would only worry about them through the `scrolled` function

Comment (crisbeto) [inline src/lib/core/overlay/scroll/scroll-dispatcher.ts]: Oh, I think I was working under the impression that this was used both for the scrollables and `scrolled`. It doesn't make sense outside of `scrolled`.

Comment (crisbeto) [inline src/lib/core/overlay/scroll/scroll-dispatcher.ts]: It's changed now.
